### PR TITLE
[JTC] Implement effort-only command interface

### PR DESCRIPTION
**Pull request by Ace314159**
This addresses the effort part of #135 and #171. I've tested this on our team's robot and it works identically to the ROS1 version from our testing.

I just realized that the trajectory tolerance parameters fix (the second commit) is also addressed by #224. I can remove the commit if you'd like to merge that first instead.

### CODE REVIEW

**Comment by Ace314159** [comment]
The test failures seem to be in `force_torque_sensor_broadcaster`, but I didn't modify that part of the code. Any ideas for what might be happening?

**Comment by destogl** [comment]
> The test failures seem to be in `force_torque_sensor_broadcaster`, but I didn't modify that part of the code. Any ideas for what might be happening?

Don't worry about that. The tests are flaky.

Could you rebase the code since we changed the formatting?

**Comment by Ace314159** [comment]
@destogl Does it look good now?

**Review by bmagyar — COMMENTED** [review]
The changes look good in general, thanks for keeping it up to date!
Do you think this should be called effort though? I think that the values you are using are basically acceleration or strongly derived from it. Does it relate to forces in any way? Effort in ros_control was used somewhat as a catch-all for "the third value" next to Position and Velocity

**Comment by Ace314159** [comment]
@bmagyar I'm not too sure either, but I just tried to mimic the behavior of ros_control.

However, I don't think this is acceleration. My team is using the effort value as the power output for the motor, so perhaps a better name might be power?

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (livanov93) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:555]: use `auto_declare` from [here](https://github.com/ros-controls/ros2_control/blob/7a0d5169c6a0f4aa8381b0e57de22ee28e6172bf/controller_interface/include/controller_interface/controller_interface.hpp#L110)

**Review by livanov93 — CHANGES_REQUESTED** [review]

**Comment by Ace314159** [comment]
@livanov93 Does this look good?

Do you know if the test failures are just caused by flaky tests or if my changes caused it? There seem to be test failures on master as well.

**Comment by livanov93** [comment]
> @livanov93 Does this look good?
> 
> Do you know if the test failures are just caused by flaky tests or if my changes caused it? There seem to be test failures on master as well.

I think it is flaky test. The rest looks fine.

**Comment by destogl** [comment]
> However, I don't think this is acceleration. My team is using the effort value as the power output for the motor, so perhaps a better name might be power?

My two cents about this. I think that you are actually using acceleration, which results in moment/power of the motor. The only thing what is intransparently happening in PID-part is conversion between acceleration and power by gain of the PID.

IMHO, the proper way for this would be to use acceleration interface and convert in robot driver acceleration to power. then if you are using other controllers you don't have to adjust gain of it for your specific robot.

(I am delighted if someone gives me a reason that this "proper" was is not so "proper") :)

**Comment by Ace314159** [comment]
@destogl Perhaps power is not the right term. In the motor code, the type is called `percent output`. The difference between what we supply to the motor and acceleration is that when we supply a value of 0, the motor completely stops, but an acceleration of 0 would still preserve the velocity. However, I don't think this is velocity either since the value we output doesn't map linearly to position

How would you convert acceleration to power? I was able to find some formulas online, but they required knowing the mass and a time for the system.

**Comment by destogl** [comment]
@Ace314159 what I can remember circular acceleration of a motor is proportional to its moment. But this discussion is probably not sensible because what you are saying you have some kind of abstract value. So we can go with the term "EFFORT" in this case. What do you think?

P.S. For personal interest, I am wondering how are you tuning your PID then if you don't have any idea about the physics of the value the motor is accepting.

**Comment by Ace314159** [comment]
@destogl Yeah, I think that sounds good.

We don't really know the physics of what the motor is accepting, but we are assuming/hoping it is consistent (which is probably not true). We can read values like the position and velocity of the motor, and we know that a positive value turns the motor in direction and a negative value turns it in the other direction. Then, we just tuned our PID following the guidelines we found from [StackExchange](https://robotics.stackexchange.com/questions/167/what-are-good-strategies-for-tuning-pid-loops). I'm not sure if a PID is the best way to do what we want, but from our cursory testing it seems to work well enough. Does this answer your question?

**Comment by theunkn0wn1** [comment]
Is there anything blocking this merge?

**Comment by Ace314159** [comment]
This is good on my end. @bmagyar @destogl Is there anything else you'd like me to do?

**Comment by theunkn0wn1** [comment]
@Ace314159 Would it be possible to get a backport of this for the Foxy branch?

**Comment by Ace314159** [comment]
@theunkn0wn1 Yes, I'll try to make a PR for that once this gets in since my team is using Foxy as well.

**Comment by Ace314159** [comment]
@bmagyar @destogl Pinging again, but is there anything blocking this merge?

**Comment by destogl** [comment]
Hi @Ace314159, we are waiting for #239 to get merged, and then we have to rebase this PR to it. This would reduce some merge collisions.

It would help us very much if you could rebase your changes to #239. Would you try this?

**Comment by Ace314159** [comment]
Thanks for the response! It looks like there are still some comments on #239, so I'll wait for those to be resolved before rebasing.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:233]: Should we use here the same logic as for velocity? To assign efforts directly if no pids are used?

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:535]: This is not needed anymore.

**Review by destogl — COMMENTED** [review]
Just a few comments.

I am a bit concerned about missing tests.

**Comment by Ace314159** [comment]
I agree. I'll try to add tests following the structure of the ones added for the velocity-only interface.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (Ace314159) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp]: Actually, would it make sense to directly set the effort or use it without PIDs? Velocity is a physical quantity, so I can see why you can set it directly. However, as we discussed and concluded before, effort is an abstract quantity, which I think means that it's not possible to directly set it. It's only used indirectly to modify the physical quantities position, velocity, and acceleration via the PID.

**Comment by Ace314159** [comment]
Do these tests look good?

I don't think the tests that seem to be failing are due to my changes since they seem to be failing on main as well.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (destogl) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp]: I don't know about that. Exactly because it is an abstract quantity, it could be “anything”. On the other hand, you are right because we would need another interpolator to see it directly. So until we have one… we can keep it also as it is.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:520]: It would be good to have also checks if command interfaces are set properly. I guess that we are currently supporting "effort" as use-only command interface.

Can you add those after line 544, similarly as done here, but without “else”. Important is to give error output if “effort” and one other command interface are defined.

Also, according to tests, it appears that the effort interface has as a requirement `"position"` and  `"velocity"` state interfaces. Can you check this also? This would help users to find faster issues with illegal configurations.

(It seems that this last check is missing for the velocity-only interface too...)

File: joint_trajectory_controller/test/test_trajectory_controller.cpp
Comment (destogl) [inline joint_trajectory_controller/test/test_trajectory_controller.cpp:1203]: It would be then great to add here checks for illegal configurations, as mentioned in the comment above.

Comment (destogl) [reply joint_trajectory_controller/test/test_trajectory_controller.cpp:1203]: The same for velocity-only mode would be nice to have.

**Review by destogl — CHANGES_REQUESTED** [review]
A few more changes! They are just minor, but let's make the code "finished" and helpful for users.

Thanks for the very fast follow-up!!!

**Comment by destogl** [comment]
> Do these tests look good?
> 
> I don't think the tests that seem to be failing are due to my changes since they seem to be failing on main as well.

This is all fine.

**Comment by Ace314159** [comment]
Thanks for the quick reviews!

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:619]: This checks here are a bit hard to understand. And it could happen that they pass if “acceleration state interface” is present. (This is also not explicitly tested.) It seems to me that there are too many checks with “else if”.

What do you think? Would it be better to separate checks? Then the code would be cleaner and functionality should be preserved, e.g.:
```suggestion
  if (has_velocity_command_interface_ && command_interface_types_.size() == 1)
```

**Review by destogl — APPROVED** [review]
Thanks for all the followups! I agree with this changes. I have only one minor comment that is potentially not covered with tests. It seems to me that there can happen error output if "acceleration" state interface is used.

Otherwise, very nice work!!!

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (Ace314159) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:619]: Just to clarify, should we allow or not allow the acceleration state interface when using effort-only or velocity-only command interfaces?

Comment (destogl) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp]: We can allow it, but I think that in this case if acceleration interface is there, other combination will not be checked. Maybe I am misunderstanding something (?).

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:613]: I think that changing this line should result in issues… Because there is no actual check for "position" and "velocity" interfaces. In the previous code, this was checked implicitly. So here should stay `else if`.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:621]: This is good now.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:630]: This should have failed on tests… because always when you are having “effort” command interface this will fail. We should at least check if there is only one interface presents, as done for the velocity. But I would be more happy if we make this checks more explicit.

They are correct, but unclear/confusing.

**Review by destogl — CHANGES_REQUESTED** [review]

**Review by destogl — APPROVED** [review]
Sorry for the last review, I was reading the code wrongly!

Everything is fine! Thanks!

**Comment by Ace314159** [comment]
@destogl Is there anything else you'd like me to do? Are we just waiting for the review from @bmagyar

**Comment by destogl** [comment]
> @destogl Is there anything else you'd like me to do? Are we just waiting for the review from @bmagyar

I am waiting for another review. I will also ping other people in today's WG meeting.

**Comment by lorepieri8** [comment]
Any news on this?

**Review by aprotyas — APPROVED** [review]

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (bmagyar) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:212]: This set of cascading `if`-s should be refactored into something more readable.
I suggest 2 things:
* make the selection of choosing the command somehow more explicit with different variable names, etc: `tmp_command_` is not very intuitive
* refactor all the `assign_interface_from_point` uses into a function that gets a parameter, the command (whether it's desired state or the pid command)

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp
Comment (bmagyar) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:602]: Investigate how to refactor these cascading `if`s so that it is easier to read which conditions together lead to a failure.
This section of  the code looks like a table of booleans where in some cases an error is flagged up.

**Review by bmagyar — APPROVED** [review]
The PR is good to go!

Two follow-up issues I created for refactoring:
https://github.com/ros-controls/ros2_controllers/issues/325
https://github.com/ros-controls/ros2_controllers/issues/326

**Comment by Ace314159** [comment]
@bmagyar  Would it be possible to backport this to foxy? Someone else also requested this and this would be useful for my team as well.

**Comment by theunkn0wn1** [comment]
Yeah, a foxy backport would be extremely useful for my team.

**Comment by destogl** [comment]
Sorry guys but that is not possible because of too many changes in the mean time. you can try to resolve conflicting PRs and then we are happy to merge. We don't have resources to do this.

**Comment by GeorgeVe** [comment]
Hee, I quikcly created a WIP backport to foxy for the joint_trajectory_controller with an effort-only command interface. It works for my team in simulation with Gazebo. It is on our forked GitLab repository: (https://gitlab.com/project-march/libraries/control/ros2_controllers/-/tree/mergify/bp/foxy/pr-225/joint_trajectory_controller).  Feel free to check it out, but **I can't guarantee that it is 100% correct and finished**. For example, the feature of 'allow_integration_in_goal_trajectories_' is also added but only partially implemented.

**Comment by rohitmenon86** [comment]
Hi, I have created a backport for galactic on my fork here https://github.com/rohitmenon86/ros2_controllers.git  for the joint_trajectory_controller ONLY with an effort-only command interface. I am still testing it

**Comment by bmagyar** [comment]
PRs are welcome

**Pull request merged**
